### ISSUE TITLE
Provide an alternative to serializeUsing to Laravel users

### ISSUE THREAD

**Issue by driesvints** [briannesbitt/Carbon#1870]
Hello,

I encountered an issue with the following code:
```php

    /**
     * @deprecated To avoid conflict between different third-party libraries, static setters should not be used.
     *             You should rather transform Carbon object before the serialization.
     *
     * JSON serialize all Carbon instances using the given callback.
     *
     * @param callable $callback
     *
     * @return void
     */
    public static function serializeUsing($callback);
```

Carbon version: **2.19.0**

PHP version: **7.3.7**

I wasn't sure where I needed to post this but I wanted to ask why this method was deprecated. We advertise this in the docs of Laravel and would like to keep using it. If this would get removed we'd probably implement this ourselves in an extended class of Carbon (which would be a bit cumbersome). Any chance that we can undeprecate this method?

Here's our docs: https://laravel.com/docs/5.8/eloquent-serialization#date-serialization

**Comment by kylekatarnls** [comment]
Hi @driesvints, glad you asked :)

It's deprecated because, `json_encode($carbonObject)` should be predictable for everyone the developer of the app and third-party libraries included.

It's closely related to:
https://github.com/laravel/framework/issues/21703
https://github.com/laravel/ideas/issues/1481

I very recommend to customize the `toResponse` to customize only `Carbon` instance in Laravel JSON outputs, and not globally in all JSON serializations.

In my humble opinion, Laravel should have it's own `$dateJsonFormat` (as it exist with `$dateFormat` for SQL) and as it exist for `toArray` customization.

This is the point of the issues above. And I think, this is the right way to go. They should be fixed first, then I would check if Carbon tooling style needs such customization for JSON output then.

> If this would get removed we'd probably implement this ourselves in an extended class of Carbon

Please don't. You would reintroduce the risk we're trying to erase.

From our side, we deprecated all static setters (`setToStringFormat`, `serializeUsing`) because they are bad patterns and make things less reliable for plugins/macros/mixins etc. which would rely on the string format they get when running `json_encode($carbonObject)` or `(string) $carbonObject`.

If you allow the app developer to customize them, then libraries can no longer use them safely and have to explicitly rewrite our methods: `json_encode($carbonObject->toISOString())` or `(string) $carbonObject->rawFormat(CarbonInterface::DEFAULT_TO_STRING_FORMAT)` that would be verbose and hassle, and most of the times, libraries developers would just miss it and calling `setToStringFormat` or `serializeUsing` would just broke those libraries.

If you want to provide a safe and  plugin-friendly framework, then you should provide encapsulated customisation over global customization. That's why we plan to do in Carbon 3.0. But don't worry, you have time. For now Carbon 3.0 is out of scope. We would let you know if we start working on it.

**Comment by driesvints** [comment]
Hey @kylekatarnls. Thanks for replying. I'll take a thorough look at this next week when I'm back from Laracon EU.

**Comment by kylekatarnls** [comment]
> Yeah would hate to see this go. We would have to implement it in our own extension class again (like we did originally). It's really useful.

I don't want this to happen. I can help to fix https://github.com/laravel/framework/issues/21703 I would send a PR if we agree on the solution.

And the deprecation can be extended until Laravel is ready to drop. But serializeUsing is a trap and it should not be deprecated.

**Comment by kylekatarnls** [comment]
Hi @taylorotwell and @driesvints, I hope you add a nice Laracon.

Did you consider the options about the usage of `serializeUsing` and how Laravel may provide JSON output customization for dates?

**Comment by driesvints** [comment]
I'll talk to taylor about this soon but this week is super busy with the Laravel 6 release and such

**Comment by austenc** [comment]
Just ran into this with an app we upgraded to Carbon 2 (and Laravel 6.0). As the Carbon upgrade guide suggests, we are adding this snippet to get the "old" behavior:
```
Carbon::serializeUsing(function ($date) {
    return [
        'date' => $date->toDateTimeString(),
        'timezone_type' => $date->timezone_type,
        'timezone' => $date->tzName,
    ];
});
```

**We're now getting an error related to `timezone_type`:**
```
Unknown getter 'timezone_type'
```

Given the discussion in this thread and deprecation of `serializeUsing`, why do the Carbon docs suggest using it? Is there a workaround in the meantime for the `timezone_type` not always being there?

**Comment by kylekatarnls** [comment]
Hi @austenc We do not recommend you to use `Carbon::serializeUsing()` we recommend you to use the default serialization (`toJSON()`). Convert a date to an ISO string is the logical thing to do as JS do the same.

So question is: why do you want dates as arrays in JSON? It could be a misuse.

Then you can simply:
```php
Carbon::serializeUsing(function ($date) {
    return [
        'date' => $date->toDateTimeString(),
        'timezone' => $date->tzName,
    ];
});
```

Unless you really need `timezone_type`, in this case please explain.

**Comment by kylekatarnls** [comment]
> why do the Carbon docs suggest using it?
The fact a method is deprecated does not mean you can't use it. It means you will have to stop using it __if__ you want to upgrade to a next major version (or the specified version). As explained above, there is no hurry, Carbon 3 is not for anytime soon. Some users create apps/sites for events that don't last and have any reason to care about deprecation. It's perfectly fine for them to mute deprecation notices and use any deprecated stuff.

But if you care about deprecation (= wanting to be ready for the future), then you should better read the reason in the notice and explained in detail above and work with new methods, that's the right choice, but it means the consumers of your JSON must handle dates as strings, so it's not a 1-line easy fix and it's not possible in every situation. So we had to mention the `Carbon::serializeUsing()` possibility in the documentation.

**Comment by austenc** [comment]
Thank you @kylekatarnls !

We got confused about the `timezone_type` mainly because it was referenced in the Carbon upgrade guide, but in some cases isn't accessible directly from the carbon object (no getter defined). You are correct about being able to omit it though! Carbon only needs the timezone to make the dates and times work correctly in our (multi-timezone) app.

When upgrading major versions of Laravel (and in this case Carbon!), our goal is to keep most things the same. In this case, we hope to (at least temporarily) use the same date serialization format that we had prior. This way we we can guarantee our app is working the same when handling dates and timezones. 

Given your explanation, we will consider `serializeUsing` to be only a temporary fix. We will update our codebase so that javascript expects dates to be serialized to strings, not arrays. Appreciate you taking the time to explain the "right" way to handle this, and for explaining the reasoning behind the decision to deprecate this method. 

Thank you once again for your response to this issue!

**Comment by kylekatarnls** [comment]
The approach I took in my apps is to dedicate a property of my model to the timezone the user chose. So if null, I fallback to user guessed/preferred timezone. In many situations it helps to have integrity. For example, if you have 2 date/time pickers for start/end of a range, you probably want the same timezone for both start and end. So having the timezone only once should be enough. And if you had a different timezone for each, you would probably be in a broken case. In this case the following exchange format would be perfectly relevant:
```json
{
  "timezone": "America/Chicago",
  "start": "2019-09-22T05:00:00.000000Z",
  "end": "2019-09-24T05:00:00.000000Z",
}
```
Then you could easily reconstitute the range in a consistent way:
```php
$json = json_decode("code above");

$start = Carbon::parse($json->start)->tz($json->timezone);
$end = Carbon::parse($json->end)->tz($json->timezone);
$range = $start->daysUntil($end);
```

In this example, the -5 hours offset is applied and you're sure both have the same timezone. And in your DB, you can save the ISO (GMT) date-time and the timezone saved in a dedicated column, so later the timezone preference can be changed without altering the date-time itself which is kept in UTC (the universal way).

This is a more detailed article about the way to work with timezone I recommend: https://medium.com/@kylekatarnls/handle-dates-the-right-way-in-php-c215650fc4fa

**Comment by driesvints** [comment]
Hey all, checking in on this again. We've already make the move to remove the method from the Laravel docs: https://github.com/laravel/docs/pull/5608

**Comment by kylekatarnls** [comment]
I think it's a good first move. Let's now determine the right approach to recommend. I think we first need a more precise customization in Laravel, something to tell Laravel how to convert `CarbonInterface` (or even `DateTimeInterface` maybe) for JSON output (including deeply nested dates).

See:
laravel/framework#21703
laravel/ideas#1481

**Comment by driesvints** [comment]
@kylekatarnls heya, I've indeed been checking into those issues as well. Are you on the Laravel Discord? Maybe it's better to chat there? Feel free to add me.

**Comment by kylekatarnls** [comment]
Hi @driesvints I'm now on the Laravel Discord Server as "kylekatarnls".

**Comment by driesvints** [comment]
@kylekatarnls heya. Discord is weird and I can't seem to find you. I'm online during most weekdays. Feel free to add me.

**Comment by kylekatarnls** [comment]
I think you're talking about a private discord channel. Let's continue by e-mail.

**Comment by kylekatarnls** [comment]
As agreed on Discord I close this issue and we'll provide a more scoped way/how-to guide to customize models dates in JSON Laravel-side.
